### PR TITLE
[REST API] Email screen for the Jetpack connection

### LINKED ISSUE #8395: [REST API] WPCom email screen for Jetpack connection
Screen Shot 2023-02-24 at 16 11 00

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #8395 

### Description
Internal docs: pe5sF9-19y-p2

This PR is the first in a series for PR to add a WordPress.com login flow for the Jetpack installation/connection, this PR has the following changes:
1. Introduce an intermediary screen to the "Jetpack Activation" flow (here I mean specifically the nav_graph of the Jetpack activation), this screen will allow choosing the start point of the flow depending on the status of site.
2. Add WordPress.com email screen.

##### WordPress.com email screen
1. Open the app, then sign in using a non-Jetpack site.
2. Click on the Jetpack benefits banner.
3. Click on "Log in to continue"
4. Confirm a WordPress.com email screen is shown.
5. Enter a non valid email, then click on Install Jetpack, confirm an error is shown.
6. Enter a non-registered email then click on Install Jetpack, confirm an error is shown.
7. Enter a valid WordPress.com email, then confirm a toast is shown with the correct event.

##### Non-regression
1. Open the app, then sign in using a Jetpack site.
2. Open the site picker.
3. Click on Add a store, then choose Connect an existing store.
4. Enter the address of a non-Jetpack site.
5. Confirm the Jetpack activation start screen is shown correctly.

### Images/gif

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComScreen.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+@Composable
+fun JetpackActivationWPComScreen(
+    viewState: JetpackActivationWPComEmailViewModel.ViewState,
+    onEmailChanged: (String) -> Unit = {},
+    onCloseClick: () -> Unit = {},
+    onContinueClick: () -> Unit = {}
+) {
+    val keyboardController = LocalSoftwareKeyboardController.current
+
+    Scaffold(
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComScreen.kt]: The layout is pretty similar to what we have in `JetpackActivationSiteCredentialsScreen`, so I copied it from it then made the necessary changes.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            onFailure = {
+                val failure = (it as? OnChangedException)?.error as? AuthOptionsError
+
+                when (failure?.type) {
+                    AuthOptionsErrorType.UNKNOWN_USER -> {
+                        errorMessage.value = R.string.email_not_registered_wpcom
+                    }
+
+                    AuthOptionsErrorType.EMAIL_LOGIN_NOT_ALLOWED -> {
+                        TODO()
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: This will be handled in a separate PR.

Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: I got a crash exactly here with my test account. Does this ever apply to regular users?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: Yes, and we will handle this before the release, that's why I left the TODO() and why I named the navigation arguments: `emailOrUsername` (sorry for not mentioning this before).

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: And it would be good to keep the test account that reproduces this at hand, it'll be useful (avoid signing in it to it by any other mean, as this removes the "suspicious" flag for 30 days).

Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: Thanks! I was mostly checking because this is being merged to trunk and I'm not sure when the TODO() work will be added in. 

I read your message too late about keeping the test account not logged-in, though. I logged in via web afterward and now the "suspicious" flag is removed and I don't get the crash anymore.

**Comment by hafizrahman** [comment]
A possible bug in the email address validation:

1. Enter an email address with space at the end, e.g: `xxxyyy@gmail.com `
2. Tap "Connect Jetpack",
3. "Enter a valid email address" error shows up,
4.  Delete the extra space at the. end, e.g `xxxyyy@gmail.com`,
5. Tap "Connect Jetpack" again,
6. "Enter a valid email address" error still shows up, despite the missing space.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComEmailViewModel.kt:79]: > Thanks! I was mostly checking because this is being merged to trunk and I'm not sure when the TODO() work will be added in.

Since this feature flagged, I think it's OK to have a TODO() in trunk 🙂.

**Comment by hichamboushaba** [comment]
Thanks @hafizrahman for the review.

> A possible bug in the email address validation:
I can't reproduce the bug with the steps you mentioned, at step 6, the login flow starts for me.

But still, I added one last commit to `trim` the email before doing the validation, I checked and this what we do in the regular email screen from the Login library, can you please re-test?

**Comment by hafizrahman** [comment]
@hichamboushaba Thanks! I was wondering why you couldn't replicate it. I found that to trigger the bug I needed to add "enter" after the space as well. So when the space was deleted, the "enter" was still actually there (although invisible), and causing the invalid error.

Doesn't matter now, though, since the latest commit fixed that 🥳

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/benefits/JetpackBenefitsDialog.kt [hunk 3]
```diff
@@ -67,14 +66,12 @@ class JetpackBenefitsDialog : DialogFragment() {
...
-                    // TODO
-                    Toast.makeText(
-                        requireContext(),
-                        "Jetpack Status: \n" +
-                            "Installed: ${event.jetpackStatus.isJetpackInstalled}\n" +
-                            "Connected: ${event.jetpackStatus.isJetpackConnected}",
-                        Toast.LENGTH_SHORT
-                    ).show()
+                    findNavController().navigateSafely(
+                        JetpackBenefitsDialogDirections.actionJetpackBenefitsDialogToJetpackActivation(
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/benefits/JetpackBenefitsDialog.kt]: Not a blocker but just wanted to discuss a bit about the usage of terms:

1. "activation"
2. "installation"

My understanding from reading the code is that "activation" is the catch-all term that covers both possibilities of "installing+connecting" and just "connecting" (in case Jetpack is already installed).

In this `setupObservers()` function, I think some usages might be confusing to incoming readers.
- the Jetpack CP case uses `installation` both for the event's name and the navigation action's name.
- the Application passwords case use `installation` for the event's name, but `activation` for the navigation action's name.


Do you think it's worth making all this consistent? If my understanding is correct, both cases above should use `activation` in all names.

**Review by hafizrahman — APPROVED** [review]
Code looks good and works well to me. Thanks for responding to my questions.

I left one more discussion, feel free to merge after looking at it.

**Comment by hafizrahman** [comment]
Maybe one other thing, that I feel could be handled in a separate PR if needed:

On landscape mode, the input field for email address is half covered when the keyboard shows up, so it's hard to see what you're typing:

Screenshot_20230228_150927

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/benefits/JetpackBenefitsDialog.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/benefits/JetpackBenefitsDialog.kt]: Good point, done in [8c07906](https://github.com/woocommerce/woocommerce-android/pull/8415/commits/8c079062cdc72b3f8438bacdad963dd0d549e5d7)

**Comment by hichamboushaba** [comment]
> On landscape mode, the input field for email address is half covered when the keyboard shows up, so it's hard to see what you're typing:

Thanks @hafizrahman, the commit [5cb3456](https://github.com/woocommerce/woocommerce-android/pull/8415/commits/5cb3456dbd8d0ae4a4a2e0ba6f37bb9e7161396a) should fix this, I changed the layout to make the whole layout scrollable.
